### PR TITLE
Switch from CircleCi to Github actions

### PR DESCRIPTION
**Pull request by lucas-shaw**
## Self-review checklist

* [ ] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [ ] (3) Tests passing
* [ ] (4) Branch ready to be merged (not work in progress)
* [ ] (5) No superfluous changes in diff
* [ ] (6) No TODO's without new ticket numbers
* [ ] (7) PR Prefixed with ticket number e.g. `CT-7654 ...`

### Related JIRA tickets
https://dsdmoj.atlassian.net/jira/software/c/projects/CDPT/boards/1152?selectedIssue=CDPT-1216

### CODE REVIEW

File: .github/workflows/test.yml
Comment (lucas-shaw) [inline .github/workflows/test.yml:1]: yarn not required for this setup although there is js, doesnt look like its used within PQ? Only asking because other services do use it

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      checks: write
+
+    env:
+      RAILS_ENV: test
+      DATABASE_URL: "postgresql://postgres:postgres@127.0.0.1/parliamentary_questions_test"
+      PQ_REST_API_HOST: http://localhost:8888
+
+    services:
+      postgres:
+        image: postgres:14-alpine
```
Comment (vertism) [inline .github/workflows/test.yml:25]: PQ is using Postgres 12

File: .github/workflows/test.yml
Comment (vertism) [reply .github/workflows/test.yml:1]: PQ doesn't have feature tests enabled at the moment. We may need to add JS dependencies as part of re-enabling them, but you can delete that stuff for now.

File: config/environments/test.rb [hunk 2]
```diff
@@ -46,7 +46,7 @@
   config.action_mailer.delivery_method = :test
   # This fills in the 'from' field which is required to run feature
   # tests on circleci
-  config.action_mailer.default_options = { from: "localhost" }
+  # config.action_mailer.default_options = { from: "localhost" }
```
Comment (vertism) [inline config/environments/test.rb:49]: What's the reason for this change?

Comment (vertism) [reply config/environments/test.rb:49]: Presumably we will need it once feature tests are turned back on (which is a ticket in our next sprint). If this is not causing an issue currently can we retain this line. You can remove the reference to CircleCI

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+                      "text": "Visit Job",
+                      "type": "button",
+                      "url": "${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"
+                    }
+                  ]
+                }
+              ]
+            }
+        env:
+          SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
```
Comment (vertism) [inline .github/workflows/deploy.yml:322]: I think this secret still needs setting up.
Also, when production is deployed, a slack message also needs to go to the release channel.
Slack config needs to be done here - https://api.slack.com/apps

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+          echo "${KUBE_CERT}" > ca.crt
+          kubectl config set-cluster ${KUBE_CLUSTER} --certificate-authority=./ca.crt --server=https://${KUBE_CLUSTER}
+          kubectl config set-credentials deploy-user --token=${KUBE_TOKEN}
+          kubectl config set-context ${KUBE_CLUSTER} --cluster=${KUBE_CLUSTER} --user=deploy-user --namespace=${KUBE_NAMESPACE}
+          kubectl config use-context ${KUBE_CLUSTER}
+
+      - name: Rollout restart deployment
+        run: |
+          kubectl set image -n ${KUBE_NAMESPACE} \
+          deployment/parliamentary-questions-development \
```
Comment (vertism) [inline .github/workflows/deploy.yml:115]: I don't think this is the correct deployment name. Look at `kubectl get deployments -n parliamentary-questions-development`

File: k8s-deploy/development/deployment_sidekiq.yaml [hunk 5]
```diff
@@ -21,6 +21,7 @@ spec:
     spec:
       containers:
         - name: parliamentary-questions-rails-jobs
+          image: 754256621582.dkr.ecr.eu-west-2.amazonaws.com/pq-team/parliamentary-questions:development.latest
```
Comment (lucas-shaw) [inline k8s-deploy/development/deployment_sidekiq.yaml:24]: Not sure whether this is required as there were not there originally. However I think its relevant to specify the image for each env

File: docker/entrypoint-webapp.sh [hunk 6]
```diff
@@ -1,14 +1,8 @@
...
 
-# Make these available via Settings in the app
-export SETTINGS__GIT_COMMIT="$APP_GIT_COMMIT"
-export SETTINGS__BUILD_DATE="$APP_BUILD_DATE"
-export SETTINGS__GIT_SOURCE="$APP_BUILD_TAG"
-
-
-rails db:migrate
-rails data:migrate
+bundle exec rails db:migrate
```
Comment (lucas-shaw) [inline docker/entrypoint-webapp.sh:4]: Imitated from the recent peoplefinder changes

File: app/services/deployment.rb
Comment (lucas-shaw) [inline app/services/deployment.rb:1]: Imitated from the recent peoplefinder changes

File: Dockerfile
Comment (lucas-shaw) [inline Dockerfile:1]: Imitated from the recent peoplefinder changes

File: .github/workflows/deploy.yml [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        run: |
+          kubectl set image -n ${KUBE_NAMESPACE} \
+          deployment/parliamentary-questions \
+          webapp="${{ vars.ECR_URL }}:$SHA"
+
+          kubectl set image -n ${KUBE_NAMESPACE} \
+          deployment/parliamentary-questions-sidekiq \
+          webapp="${{ vars.ECR_URL }}:$SHA"
+
+      - name: Send deploy notification to product Slack channel
```
Comment (vertism) [inline .github/workflows/deploy.yml:307]: The slack notification for production needs to go to the product channel and the production deploys channel. You'll need to duplicate the code to do it using a different webhook url.

This also needs doing on Peoplefinder btw.

File: .github/workflows/deploy.yml [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+          KUBE_TOKEN: ${{ secrets.KUBE_TOKEN }}
+          KUBE_CLUSTER: ${{ secrets.KUBE_CLUSTER }}
+        run: |
+          echo "${KUBE_CERT}" > ca.crt
+          kubectl config set-cluster ${KUBE_CLUSTER} --certificate-authority=./ca.crt --server=https://${KUBE_CLUSTER}
+          kubectl config set-credentials deploy-user --token=${KUBE_TOKEN}
+          kubectl config set-context ${KUBE_CLUSTER} --cluster=${KUBE_CLUSTER} --user=deploy-user --namespace=${KUBE_NAMESPACE}
+          kubectl config use-context ${KUBE_CLUSTER}
+
+      - name: Rollout restart deployment
```
Comment (vertism) [inline .github/workflows/deploy.yml:112]: We need to recreate what happens in https://github.com/ministryofjustice/parliamentary-questions/blob/main/deploy.sh, so that will include updating the image on cronjobs.

File: .github/workflows/deploy.yml [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+
+          kubectl set image -n ${KUBE_NAMESPACE} \
+          jobs/trim-database \
+          webapp="${{ vars.ECR_URL }}:$SHA"
+
+          kubectl set image -n ${KUBE_NAMESPACE} \
+          jobs/early-bird-dispatch \
+          webapp="${{ vars.ECR_URL }}:$SHA"
+
+          kubectl set image -n ${KUBE_NAMESPACE} \
```
[the remark this answers is not in the document]
Comment (vertism) [reply .github/workflows/deploy.yml:230]: Looks like they are. I would put them first so that the migration should have run by the time the new webapp containers are deployed.

File: .github/workflows/deploy.yml [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+          jobs/nightly-import \
+          webapp="${{ vars.ECR_URL }}:$SHA"
+
+          kubectl set image -n ${KUBE_NAMESPACE} \
+          jobs/trim-database \
+          webapp="${{ vars.ECR_URL }}:$SHA"
+
+          kubectl set image -n ${KUBE_NAMESPACE} \
```
Comment (vertism) [inline .github/workflows/deploy.yml:226]: This isn't run in staging

File: docker/entrypoint-webapp.sh (hunk 6 above)
Comment (vertism) [inline docker/entrypoint-webapp.sh:4]: Seeing as the migration is done here, I don't think the rails-migrations jobs are needed, so probably should be removed the deploy instructions.
